images/bb/cmd/bb/main.go

```diff
@@ -0,0 +1,125 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+	"sync"
+
+	"cuelang.org/go/cue"
+	"cuelang.org/go/cue/cuecontext"
+	cueerrors "cuelang.org/go/cue/errors"
+	"cuelang.org/go/cue/load"
+	"tractor.dev/toolkit-go/engine"
+	"tractor.dev/toolkit-go/engine/cli"
+	"tractor.dev/toolkit-go/engine/daemon"
+
+	"github.com/ajbouh/substrate/images/bb/blackboard"
+	"github.com/ajbouh/substrate/images/bb/calldef"
+	bbhttp "github.com/ajbouh/substrate/images/bb/http"
+	"github.com/ajbouh/substrate/pkg/cueloader"
+)
+
+type Main struct {
+	listenAddr string
+
+	Daemon *daemon.Framework
+}
+
+func main() {
+	engine.Run(
+		Main{
+			listenAddr: ":" + os.Getenv("PORT"),
+		},
+	)
+}
+
+func (m *Main) InitializeCLI(root *cli.Command) {
+	// a workaround for an unresolved issue in toolkit-go/engine
+	// for figuring out if its a CLI or a daemon program...
+	root.Run = func(ctx *cli.Context, args []string) {
+		if err := m.Daemon.Run(ctx); err != nil {
+			log.Fatal(err)
+		}
+	}
+}
+
+func (m *Main) Serve(ctx context.Context) {
+	cc := cuecontext.New()
+	cueMu := &sync.Mutex{}
+
+	callDefLoader := calldef.NewLoader(
+		cueloader.NewCueLoader(
+			":defs",
+			cueloader.LookupPathTransform(cue.MakePath(cue.Def("#out"), cue.Def("#lenses"))),
+		),
+	)
+
+	mu := &sync.RWMutex{}
+	var bb *blackboard.Blackboard
+
+	// close when ready
+	ready := make(chan struct{})
+	markReady := func() { close(ready) }
+	var markReadyOnce sync.Once
```

Comment:
I'll sometimes use a context to listen on `<-Done()` to simplify this a bit:
```
readyCtx, markReady := context.WithCancel(context.Background())
```

Reply:
Yes. Better to limp forward than hang, I think.